api/routes.go

```diff
@@ -0,0 +1,194 @@
+		router.POST("/wallet/init", RequirePassword(api.walletInitHandler, requiredPassword))
+		router.POST("/wallet/init/seed", RequirePassword(api.walletInitSeedHandler, requiredPassword))
+		router.POST("/wallet/lock", RequirePassword(api.walletLockHandler, requiredPassword))
+		router.POST("/wallet/seed", RequirePassword(api.walletSeedHandler, requiredPassword))
+		router.GET("/wallet/seeds", RequirePassword(api.walletSeedsHandler, requiredPassword))
+		router.POST("/wallet/siacoins", RequirePassword(api.walletSiacoinsHandler, requiredPassword))
+		router.POST("/wallet/siafunds", RequirePassword(api.walletSiafundsHandler, requiredPassword))
+		router.POST("/wallet/siagkey", RequirePassword(api.walletSiagkeyHandler, requiredPassword))
+		router.POST("/wallet/sweep/seed", RequirePassword(api.walletSweepSeedHandler, requiredPassword))
+		router.GET("/wallet/transaction/:id", api.walletTransactionHandler)
+		router.GET("/wallet/transactions", api.walletTransactionsHandler)
+		router.GET("/wallet/transactions/:addr", api.walletTransactionsAddrHandler)
+		router.GET("/wallet/verify/address/:addr", api.walletVerifyAddressHandler)
+		router.POST("/wallet/unlock", RequirePassword(api.walletUnlockHandler, requiredPassword))
+		router.POST("/wallet/changepassword", RequirePassword(api.walletChangePasswordHandler, requiredPassword))
+	}
+
+	// Apply UserAgent middleware and return the Router
+	api.router = cleanCloseHandler(RequireUserAgent(router, requiredUserAgent))
+	return router
+}
+
+// RequireUserAgent is middleware that requires all requests to set a
+// UserAgent that contains the specified string.
+func RequireUserAgent(h http.Handler, ua string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if !strings.Contains(req.UserAgent(), ua) {
+			WriteError(w, Error{"Browser access disabled due to security vulnerability. Use Sia-UI or siac."}, http.StatusBadRequest)
+			return
+		}
+		h.ServeHTTP(w, req)
+	})
+}
+
+// RequirePassword is middleware that requires a request to authenticate with a
+// password using HTTP basic auth. Usernames are ignored. Empty passwords
+// indicate no authentication is required.
+func RequirePassword(h httprouter.Handle, password string) httprouter.Handle {
+	// An empty password is equivalent to no password.
+	if password == "" {
+		return h
+	}
+	return func(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+		_, pass, ok := req.BasicAuth()
+		if !ok || pass != password {
+			w.Header().Set("WWW-Authenticate", "Basic realm=\"SiaAPI\"")
+			WriteError(w, Error{"API authentication failed."}, http.StatusUnauthorized)
+			return
+		}
+		h(w, req, ps)
+	}
+}
+
+// cleanCloseHandler wraps the entire API, ensuring that underlying conns are
+// not leaked if the rmeote end closes the connection before the underlying
+// handler finishes.
+func cleanCloseHandler(next http.Handler) http.Handler {
```

Comment:
Should this be defined in `routes.go`? Seem orthogonal to route definition, as do some of the other funcs (`RequireUserAgent` etc)

Reply:
i don't mind wherever it goes and it seems like it should stick with the `api.router = ` piece @avahowell